iroha_network/src/lib.rs

```diff
@@ -56,13 +56,15 @@ impl Network {
     ///
     /// * `server_url` - url of format ip:port (e.g. `127.0.0.1:7878`) on which this server will listen for incoming connections.
     /// * `handler` - callback function which is called when there is an incoming connection, it get's the stream for this connection
-    pub fn listen<H>(server_url: &str, mut handler: H) -> Result<(), Box<dyn Error>>
+    /// * `state` - the state that you want to capture, place `()` there if don't need the state, use some reference to the state otherwise
+    pub fn listen<H, S>(state: S, server_url: &str, mut handler: H) -> Result<(), String>
     where
-        H: FnMut(Box<dyn Stream>) -> Result<(), Box<dyn Error>>,
+        H: FnMut(S, Box<dyn Stream>) -> Result<(), String>,
+        S: Clone,
     {
-        let listener = TcpListener::bind(server_url)?;
+        let listener = TcpListener::bind(server_url).map_err(|e| e.to_string())?;
         while let Some(stream) = listener.incoming().next() {
-            handler(Box::new(stream?))?;
+            handler(state.clone(), Box::new(stream.map_err(|e| e.to_string())?))?;
```

Comment:
The purpose to do not clone, but sync this state between handlers.
Common instructions queue should be available to processor and all handlers, with cloning it will just spawn queues.

Reply:
For example it can be shared via mutex: https://docs.rs/actix-web/3.0.0-alpha.1/actix_web/web/struct.Data.html

Reply:
You can take a look on `single_threaded_async_stateful` example where I show how to use it. Basically it should be a reference and the reference will be cloned not content.

Reply:
I tried not to make any assumption on the type, so any references and unit type can be used there.